database/internal/regexcache/list.go

```diff
@@ -0,0 +1,90 @@
+package regexcache
+
+type regexElement struct {
+	next, prev *regexElement
```

Comment:
FWIW this looks like doubly linked list to me, which is already available in Go standard library (see [container/list](https://golang.org/pkg/container/list)).

Reply:
Type-safety is single reason. 
In general, we can just use https://github.com/hashicorp/golang-lru or any another cache library.